src/Workspaces/Core/Portable/Editing/ImportAdderService.cs

```diff
@@ -18,44 +20,240 @@ namespace Microsoft.CodeAnalysis.Editing
-            bool isInSpan(SyntaxNodeOrToken nodeOrToken) =>
-                spansTree.HasIntervalThatOverlapsWith(nodeOrToken.FullSpan.Start, nodeOrToken.FullSpan.Length);
+            bool isInSpan(SyntaxNode node) =>
+                spansTree.HasIntervalThatOverlapsWith(node.FullSpan.Start, node.FullSpan.Length);
 
-            var nodesWithExplicitNamespaces = root.DescendantNodesAndSelf().Where(n => isInSpan(n) && GetExplicitNamespaceSymbol(n, model) != null).ToList();
+            var nodes = root.DescendantNodesAndSelf().Where(isInSpan);
+            var (importDirectivesToAdd, context) = strategy switch
+            {
+                Strategy.AddImportsFromSymbolAnnotations
+                    => GetImportDirectivesFromAnnotatedNodesAsync(nodes, root, model, addImportsService, generator, cancellationToken),
+                Strategy.AddImportsFromSyntaxes
+                    => GetImportDirectivesFromSyntaxesAsync(nodes, ref root, model, addImportsService, generator, cancellationToken),
+                _ => throw new InvalidEnumArgumentException(nameof(strategy), (int)strategy, typeof(Strategy)),
+            };
 
-            var namespacesToAdd = new HashSet<INamespaceSymbol>();
-            namespacesToAdd.AddRange(nodesWithExplicitNamespaces.Select(
-                n => GetExplicitNamespaceSymbol(n, model)));
+            var placeSystemNamespaceFirst = options.GetOption(GenerationOptions.PlaceSystemNamespaceFirst, document.Project.Language);
 
-            var generator = SyntaxGenerator.GetGenerator(document);
-            var imports = namespacesToAdd.Select(ns => generator.NamespaceImportDeclaration(ns.ToDisplayString()).WithAdditionalAnnotations(Simplifier.Annotation))
-                                         .ToArray();
+            root = addImportsService.AddImports(model.Compilation, root, context, importDirectivesToAdd, placeSystemNamespaceFirst);
 
-            // annotate these nodes so they get simplified later
-            var newRoot = root.ReplaceNodes(
-                nodesWithExplicitNamespaces,
-                (o, r) => r.WithAdditionalAnnotations(Simplifier.Annotation));
+            return document.WithSyntaxRoot(root);
+        }
 
-            var placeSystemNamespaceFirst = options.GetOption(GenerationOptions.PlaceSystemNamespaceFirst, document.Project.Language);
-            var addImportsService = document.GetLanguageService<IAddImportsService>();
-            var finalRoot = addImportsService.AddImports(
-                model.Compilation, newRoot, newRoot, imports, placeSystemNamespaceFirst);
+        protected abstract INamespaceSymbol GetExplicitNamespaceSymbol(SyntaxNode node, SemanticModel model);
 
-            return document.WithSyntaxRoot(finalRoot);
+        /// <summary>
+        /// Gets the namespace <paramref name="node"/> is contained in, or null otherwise
+        /// </summary>
+        protected abstract INamespaceSymbol GetContainedNamespace(SyntaxNode node, SemanticModel model);
```

Comment:
so we already have a SemanticModelExtension.GetEnclosingNamespace.  Will that not work here?